# Plotting with Seaborn
## Milestone 1: Getting Data Ready

### Objective
In this milestone, you will read the dataset into a pandas DataFrame and perform some data wrangling on it to make it more suitable for better visualization.
### Importance of the Project
The step of data wrangling is an important part of almost all data analysis, data science, and machine learning projects. This task ensures that further analysis of the data yields the answers you are interested in finding out.

### Dataset
- **age**: Age in years (continuous)
- **workclass**: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- **fnlwgt**: (continuous) Represents final weight, which is the number of units in the target population that the responding unit represents. 
- **education**: 
    - Bachelors
    - Some-college
    - 11th
    - HS-grad
    - Prof-school
    - Assoc-acdm
    - Assoc-voc
    - 9th, 7th-8th, 12th
    - Masters
    - 1st-4th, 10th
    - Doctorate
    - 5th-6th
    - Preschool.
- **education-num**: Number of years of education (continuous). Stands for the number of years of education in total, which is a continuous representation of the discrete variable education. 
- **marital-status**: 
    - Married-civ-spouse
    - Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- **occupation**: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- **relationship**: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried. Represents the responding unit’s role in the family.
- **race**: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- **sex**: Female, Male.
- **capital-gain**: (continuous) Income from investment sources other than wage/salary.
- **capital-loss**: (continuous) Income from investment sources other than wage/salary.
- **hours-per-week**: Hours worked per week (continuous)
- **native-country**: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

### 1. Import the required Python packages.

In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.set_option('display.max_rows', 100)

### 2. Read the dataset into a pandas DataFrame.
read adult_consensus_data.csv

In [3]:
data_loc = "./data/"

In [4]:
df = pd.read_csv(data_loc + "adult_consensus_data.csv")

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       32561 non-null  int64 
 1   workclass                 32561 non-null  object
 2   fnlwgt                    32561 non-null  int64 
 3   education                 32561 non-null  object
 4   education-num             32561 non-null  int64 
 5   marital-status            32561 non-null  object
 6   occupation                32561 non-null  object
 7   relationship              32561 non-null  object
 8   race                      32561 non-null  object
 9   sex                       32561 non-null  object
 10  capital-gain              32561 non-null  int64 
 11  capital-loss              32561 non-null  int64 
 12  hours-per-week            32561 non-null  int64 
 13  native-country            32561 non-null  object
 14  discretized_gross_inco

In [6]:
df.shape

(32561, 15)

In [7]:
df.columns

Index(['Age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'discretized_gross_income'],
      dtype='object')

In [8]:
# normalize the column names: take everything to lowercase and replace dashes with underscore
df.rename(columns = lambda x: x.strip().lower().replace("-","_"), inplace = True)

#### **Function:** `DataFrame.rename`

The method **`DataFrame.rename()`** changes the labels of columns or rows in a DataFrame. Here it is used with a **function** instead of a dictionary, meaning it applies the same transformation to *all* column names.

---

##### **Code Example**

```python
df.rename(
    columns = lambda x: x.strip().lower().replace("-", "_"), 
    inplace = True
)
```

---

##### **Step-by-Step Explanation**

1. **`columns=`**

   * Tells Pandas you want to rename the **columns** (not the row index).

2. **`lambda x: ...`**

   * Instead of mapping old → new names manually, we pass a **function** (`lambda`) that is applied to each column name (`x`).
   * This is useful when you want to apply the same cleaning rule to all names.

3. **`x.strip()`**

   * Removes any leading or trailing spaces in the column name.
   * Example: `" Temp "` → `"Temp"`.

4. **`.lower()`**

   * Converts the cleaned name to lowercase.
   * Example: `"Temp"` → `"temp"`.

5. **`.replace("-", "_")`**

   * Replaces any hyphens (`-`) with underscores (`_`).
   * Example: `"temp-value"` → `"temp_value"`.

6. **`inplace=True`**

   * Applies the changes directly to the existing DataFrame `df`.
   * Without this, Pandas would return a *new* DataFrame instead.

---

##### **Result**

All column names in `df` will now:

* Have no extra spaces,
* Be lowercase,
* Use underscores instead of hyphens.

Example transformation:

```
" Temp-Value "   →   "temp_value"
```

---

**In short:**
This line standardizes column names in the DataFrame, making them clean, consistent, and easier to work with in code.

In [9]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'discretized_gross_income'],
      dtype='object')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   age                       32561 non-null  int64 
 1   workclass                 32561 non-null  object
 2   fnlwgt                    32561 non-null  int64 
 3   education                 32561 non-null  object
 4   education_num             32561 non-null  int64 
 5   marital_status            32561 non-null  object
 6   occupation                32561 non-null  object
 7   relationship              32561 non-null  object
 8   race                      32561 non-null  object
 9   sex                       32561 non-null  object
 10  capital_gain              32561 non-null  int64 
 11  capital_loss              32561 non-null  int64 
 12  hours_per_week            32561 non-null  int64 
 13  native_country            32561 non-null  object
 14  discretized_gross_inco

In [11]:
df.head(10)

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,discretized_gross_income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
5,37,Private,284582,Masters,14,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,<=50K
6,49,Private,160187,9th,5,Married-spouse-absent,Other-service,Not-in-family,Black,Female,0,0,16,Jamaica,<=50K
7,52,Self-emp-not-inc,209642,HS-grad,9,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,45,United-States,>50K
8,31,Private,45781,Masters,14,Never-married,Prof-specialty,Not-in-family,White,Female,14084,0,50,United-States,>50K
9,42,Private,159449,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,5178,0,40,United-States,>50K


In [12]:
df.sample(5)

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,discretized_gross_income
29519,38,Private,194304,Some-college,10,Divorced,Transport-moving,Not-in-family,Black,Male,0,0,55,United-States,<=50K
5397,23,Self-emp-not-inc,258298,Bachelors,13,Never-married,Adm-clerical,Own-child,White,Male,0,2231,40,United-States,>50K
7932,29,Private,130620,Some-college,10,Married-spouse-absent,Sales,Own-child,Asian-Pac-Islander,Female,0,0,26,India,<=50K
30675,44,Private,155472,Assoc-acdm,12,Never-married,Prof-specialty,Unmarried,Black,Female,1151,0,50,United-States,<=50K
23384,31,Federal-gov,166626,HS-grad,9,Married-civ-spouse,Adm-clerical,Husband,White,Male,0,0,40,United-States,>50K


#### **Function:** `DataFrame.sample`

The **`DataFrame.sample()`** method returns a random subset of rows from a DataFrame. It’s often used for quick inspection, debugging, or creating smaller samples for testing or exploration.

---

##### **Code Example**

```python
df.sample(5)
```

---

##### **Step-by-Step Explanation**

1. **`df`**

   * Your DataFrame (with rows and columns of data).

2. **`.sample(5)`**

   * Requests a random selection of 5 rows from the DataFrame.
   * Each run may return different rows unless you set a `random_state` (seed).

---

##### **Optional Parameters**

* **`random_state`**: Fixes the seed so you always get the same sample.

  ```python
  df.sample(5, random_state=42)
  ```
* **`frac`**: Instead of a fixed number, choose a fraction of rows.

  ```python
  df.sample(frac=0.1)   # returns 10% of the DataFrame
  ```
* **`axis`**: By default samples rows (`axis=0`), but you can sample columns (`axis=1`).

  ```python
  df.sample(3, axis=1)   # pick 3 random columns
  ```

---

##### **Result**

`df.sample(5)` returns 5 random rows from the DataFrame, with all their columns.
It doesn’t change the original DataFrame — it just creates a new, smaller view.

---

**In short:**
`df.sample(5)` gives you a quick, random snapshot of 5 rows from your DataFrame. Perfect for data exploration when `.head()` or `.tail()` feels too limited.

In [13]:
df[0:5]

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,discretized_gross_income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


### 3. Perform basic exploration of data.
- Use utilities likeshape, columns, head(), and unique() to find the features of the dataset.
- Look for missing values and unusual values such ? during this exploratory step.
- Observe that data values in categorical columns have leading spaces.

#### Workclass

In [14]:
# Check unique values of workclass using unique 
df.workclass.unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay',
       ' Never-worked'], dtype=object)

#### **Function:** `Series.unique`

The **`unique()`** method returns the distinct values in a Pandas Series (one column of a DataFrame).

---

##### **Code Example**

```python
df.workclass.unique()
```

---

##### **Step-by-Step Explanation**

1. **`df.workclass`**

   * Accesses the **`workclass`** column in the DataFrame `df`.
   * This returns a Pandas Series (a 1D array of values).

2. **`.unique()`**

   * Scans through all the values in the column.
   * Collects each **unique (non-duplicate) entry**.
   * Returns them as a NumPy array.

---

##### **Result**

```
array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay',
       ' Never-worked'], dtype=object)
```

* This tells you all the different categories of workclass in the dataset:

  * `"State-gov"`
  * `"Self-emp-not-inc"`
  * `"Private"`
  * `"Federal-gov"`
  * `"Local-gov"`
  * `"?"` (probably missing or unknown values)
  * `"Self-emp-inc"`
  * `"Without-pay"`
  * `"Never-worked"`
* `dtype=object` means the data type is strings (object type in Pandas/NumPy).

---

##### **Why it’s useful**

* Quickly shows you the categories present in a column (common for categorical variables).
* Helps detect issues like:

  * Leading/trailing spaces (notice `' State-gov'` has a space at the start!).
  * Unexpected values (`' ?'` looks like a placeholder for missing data).

---

**In short:**
`df.workclass.unique()` lists all the distinct categories in the *workclass* column. It’s a handy first step in exploring categorical data and spotting cleaning tasks.

In [15]:
# Check unique values of workclass using groupby 
sel_column = "workclass"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
workclass,
?,1836
Federal-gov,960
Local-gov,2093
Never-worked,7
Private,22696
Self-emp-inc,1116
Self-emp-not-inc,2541
State-gov,1298
Without-pay,14


#### **Counting with groupby**

```python
# Check unique values of workclass using groupby 
sel_column = "workclass"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()
```

---

##### **Step-by-Step Explanation**

1. **`sel_column = "workclass"`**

   * Stores the column name `"workclass"` in a variable.
   * This makes the code flexible — if you change `sel_column`, the rest of the code updates automatically.

---

2. **`df.groupby(sel_column)`**

   * Groups the DataFrame `df` by the values in the `workclass` column.
   * Each unique value of `workclass` (e.g., `"Private"`, `"State-gov"`, `"?"`) becomes a group.

---

3. **`.size()`**

   * Counts how many rows fall into each group.
   * Unlike `.count()`, this counts all rows, regardless of whether other columns contain `NaN`.
   * Example: If `"Private"` appears in 22,696 rows, you get `22696`.

---

4. **`.to_frame(name="row_count")`**

   * Converts the result (which is a Pandas Series) into a DataFrame.
   * Names the column `"row_count"` instead of the default `0`.
   * This makes the output table clearer.

---

5. **`.sort_index()`**

   * Sorts the resulting DataFrame by the group labels (the `workclass` categories).
   * Ensures the rows are ordered alphabetically (or numerically, depending on index type).

---

##### **Why this is useful**

* Combines category inspection (`unique()`) with row counts (`size()`) into one step.
* The table shows not just which categories exist, but how many observations each has.
* Makes it easier to spot imbalances (e.g., `"Private"` dominates) or rare categories (`"Never-worked"`, `"Without-pay"`).

---

**In short:**
This snippet groups the data by `workclass`, counts how many rows belong to each category, and outputs a neat DataFrame with those counts, sorted by the category name.

#### Education

In [16]:
# Check unique values of education using unique
df.education.unique()

array([' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th',
       ' Some-college', ' Assoc-acdm', ' Assoc-voc', ' 7th-8th',
       ' Doctorate', ' Prof-school', ' 5th-6th', ' 10th', ' 1st-4th',
       ' Preschool', ' 12th'], dtype=object)

In [17]:
# Check unique values of education using groupby
sel_column = "education"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
education,
10th,933
11th,1175
12th,433
1st-4th,168
5th-6th,333
7th-8th,646
9th,514
Assoc-acdm,1067
Assoc-voc,1382


#### Education Num

In [18]:
# Check unique values of education_num using unique
df.education_num.unique()

array([13,  9,  7, 14,  5, 10, 12, 11,  4, 16, 15,  3,  6,  2,  1,  8])

In [19]:
# Check unique values of education_num using groupby
sel_column = "education_num"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
education_num,
1,51
2,168
3,333
4,646
5,514
6,933
7,1175
8,433
9,10501


#### Marital Status

In [20]:
# Check unique values of marital_status using unique
df.marital_status.unique()

array([' Never-married', ' Married-civ-spouse', ' Divorced',
       ' Married-spouse-absent', ' Separated', ' Married-AF-spouse',
       ' Widowed'], dtype=object)

In [21]:
# Check unique values of marital_status using groupby
sel_column = "marital_status"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
marital_status,
Divorced,4443
Married-AF-spouse,23
Married-civ-spouse,14976
Married-spouse-absent,418
Never-married,10683
Separated,1025
Widowed,993


#### Occupation

In [22]:
# Check unique values of occupation using unique
df.occupation.unique()

array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Sales', ' Craft-repair',
       ' Transport-moving', ' Farming-fishing', ' Machine-op-inspct',
       ' Tech-support', ' ?', ' Protective-serv', ' Armed-Forces',
       ' Priv-house-serv'], dtype=object)

In [23]:
# Check unique values of occupation using groupby
sel_column = "occupation"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
occupation,
?,1843
Adm-clerical,3770
Armed-Forces,9
Craft-repair,4099
Exec-managerial,4066
Farming-fishing,994
Handlers-cleaners,1370
Machine-op-inspct,2002
Other-service,3295


#### Sex

In [24]:
# Check unique values of sex using unique
df.sex.unique()

array([' Male', ' Female'], dtype=object)

In [25]:
# Check unique values of sex using groupby
sel_column = "sex"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
sex,
Female,10771
Male,21790


#### Native Country

In [26]:
# Check unique values of native_country using unique
df.native_country.unique()

array([' United-States', ' Cuba', ' Jamaica', ' India', ' ?', ' Mexico',
       ' South', ' Puerto-Rico', ' Honduras', ' England', ' Canada',
       ' Germany', ' Iran', ' Philippines', ' Italy', ' Poland',
       ' Columbia', ' Cambodia', ' Thailand', ' Ecuador', ' Laos',
       ' Taiwan', ' Haiti', ' Portugal', ' Dominican-Republic',
       ' El-Salvador', ' France', ' Guatemala', ' China', ' Japan',
       ' Yugoslavia', ' Peru', ' Outlying-US(Guam-USVI-etc)', ' Scotland',
       ' Trinadad&Tobago', ' Greece', ' Nicaragua', ' Vietnam', ' Hong',
       ' Ireland', ' Hungary', ' Holand-Netherlands'], dtype=object)

In [27]:
# Check unique values of native_country using groupby
sel_column = "native_country"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
native_country,
?,583
Cambodia,19
Canada,121
China,75
Columbia,59
Cuba,95
Dominican-Republic,70
Ecuador,28
El-Salvador,106


#### Discretized Gross Income

In [28]:
# Check unique values of discretized_gross_income using unique
df.discretized_gross_income.unique()

array([' <=50K', ' >50K'], dtype=object)

In [29]:
# Check unique values of discretized_gross_income using groupby
sel_column = "discretized_gross_income"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
discretized_gross_income,
<=50K,24720
>50K,7841


#### Capital Gain

In [30]:
# Check unique values of capital_gain using unique
df.capital_gain.unique()

array([ 2174,     0, 14084,  5178,  5013,  2407, 14344, 15024,  7688,
       34095,  4064,  4386,  7298,  1409,  3674,  1055,  3464,  2050,
        2176,   594, 20051,  6849,  4101,  1111,  8614,  3411,  2597,
       25236,  4650,  9386,  2463,  3103, 10605,  2964,  3325,  2580,
        3471,  4865, 99999,  6514,  1471,  2329,  2105,  2885, 25124,
       10520,  2202,  2961, 27828,  6767,  2228,  1506, 13550,  2635,
        5556,  4787,  3781,  3137,  3818,  3942,   914,   401,  2829,
        2977,  4934,  2062,  2354,  5455, 15020,  1424,  3273, 22040,
        4416,  3908, 10566,   991,  4931,  1086,  7430,  6497,   114,
        7896,  2346,  3418,  3432,  2907,  1151,  2414,  2290, 15831,
       41310,  4508,  2538,  3456,  6418,  1848,  3887,  5721,  9562,
        1455,  2036,  1831, 11678,  2936,  2993,  7443,  6360,  1797,
        1173,  4687,  6723,  2009,  6097,  2653,  1639, 18481,  7978,
        2387,  5060])

In [31]:
# Check unique values of capital_gain using groupby
sel_column = "capital_gain"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
capital_gain,
0,29849
114,6
401,2
594,34
914,8
...,...
25236,11
27828,34
34095,5


#### Capital Loss

In [32]:
# Check unique values of capital_loss using unique
df.capital_loss.unique()

array([   0, 2042, 1408, 1902, 1573, 1887, 1719, 1762, 1564, 2179, 1816,
       1980, 1977, 1876, 1340, 2206, 1741, 1485, 2339, 2415, 1380, 1721,
       2051, 2377, 1669, 2352, 1672,  653, 2392, 1504, 2001, 1590, 1651,
       1628, 1848, 1740, 2002, 1579, 2258, 1602,  419, 2547, 2174, 2205,
       1726, 2444, 1138, 2238,  625,  213, 1539,  880, 1668, 1092, 1594,
       3004, 2231, 1844,  810, 2824, 2559, 2057, 1974,  974, 2149, 1825,
       1735, 1258, 2129, 2603, 2282,  323, 4356, 2246, 1617, 1648, 2489,
       3770, 1755, 3683, 2267, 2080, 2457,  155, 3900, 2201, 1944, 2467,
       2163, 2754, 2472, 1411])

In [33]:
# Check unique values of capital_loss using groupby
sel_column = "capital_loss"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
capital_loss,
0,31042
155,1
213,4
323,3
419,3
625,12
653,3
810,2
880,6


#### Hours per Week

In [34]:
# Check unique values of hours_per_week using unique
df.hours_per_week.unique()

array([40, 13, 16, 45, 50, 80, 30, 35, 60, 20, 52, 44, 15, 25, 38, 43, 55,
       48, 58, 32, 70,  2, 22, 56, 41, 28, 36, 24, 46, 42, 12, 65,  1, 10,
       34, 75, 98, 33, 54,  8,  6, 64, 19, 18, 72,  5,  9, 47, 37, 21, 26,
       14,  4, 59,  7, 99, 53, 39, 62, 57, 78, 90, 66, 11, 49, 84,  3, 17,
       68, 27, 85, 31, 51, 77, 63, 23, 87, 88, 73, 89, 97, 94, 29, 96, 67,
       82, 86, 91, 81, 76, 92, 61, 74, 95])

In [35]:
# Check unique values of hours_per_week using groupby
sel_column = "hours_per_week"
df.groupby(sel_column).size().to_frame(name="row_count").sort_index()

,row_count
hours_per_week,
1,20
2,32
3,39
4,54
5,60
6,64
7,26
8,145
9,18


### 4. Drop the columns fnlwgt, race, and relationship from the DataFrame.

In [36]:
df.drop(labels = ["fnlwgt", "race", "relationship"], axis = 1, inplace = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   age                       32561 non-null  int64 
 1   workclass                 32561 non-null  object
 2   education                 32561 non-null  object
 3   education_num             32561 non-null  int64 
 4   marital_status            32561 non-null  object
 5   occupation                32561 non-null  object
 6   sex                       32561 non-null  object
 7   capital_gain              32561 non-null  int64 
 8   capital_loss              32561 non-null  int64 
 9   hours_per_week            32561 non-null  int64 
 10  native_country            32561 non-null  object
 11  discretized_gross_income  32561 non-null  object
dtypes: int64(5), object(7)
memory usage: 3.0+ MB


#### **Code Explanation**

1. **`df.drop(...)`**

   * `drop()` removes rows or columns from a DataFrame.
   * **`labels = ["fnlwgt", "race", "relationship"]`**: tells pandas which columns you want to remove.
   * **`axis = 1`**: specifies that you’re dropping columns (if it were `axis = 0`, it would drop rows instead).
   * **`inplace = True`**: means the changes are applied directly to `df` (the DataFrame is modified in memory). Without this, you’d get back a new DataFrame with those columns dropped, and `df` would remain unchanged.

   After this line, the columns `"fnlwgt"`, `"race"`, and `"relationship"` are permanently removed from `df`.

---

2. **`df.info()`**

   * Prints a summary of the DataFrame structure.
   * Shows:

     * Number of entries (`32561`) and the index range (`0` to `32560`).
     * Total number of columns (`12` now, because 3 were dropped).
     * Column names, non-null counts, and data types (`int64` for numbers, `object` for text).
     * Memory usage (about 3.0 MB).

---

##### **Result**

Before dropping, the DataFrame had 15 columns. After dropping, it has 12 columns, as confirmed by `df.info()`.

Remaining columns include:

* **Numerical variables**: `age`, `education_num`, `capital_gain`, `capital_loss`, `hours_per_week`.
* **Categorical variables**: `workclass`, `education`, `marital_status`, `occupation`, `sex`, `native_country`, `discretized_gross_income`.

---

##### **Why this is useful**

* Cleaning/preparing a dataset often involves removing unnecessary or redundant columns.
* For example, `fnlwgt` (final sampling weight) is often dropped in the Adult Income dataset because it isn’t useful for prediction tasks.
* `race` and `relationship` might be excluded for privacy or relevance reasons, depending on the analysis goal.

---

**In short:**
This code removes three unneeded columns from the DataFrame (`fnlwgt`, `race`, `relationship`) and then uses `.info()` to confirm the DataFrame now has 12 columns and no missing values.

### 5. Handle missing values.
In Step 3, when you apply the function unique() to the columns of the DataFrame you observe that in columns 'workclass’, 'occupation’, and ‘native_country’ there are missing values represented by ? in the DataFrame. 
Drop all the rows containing ? in one or more columns.

In [37]:
from pandas.api.types import is_object_dtype

In [38]:
# is_object_dtype checks whether the Series (or array) has an object dtype.
# Returns True if the column is object-based (often strings, sometimes mixed types).
# Returns False otherwise (e.g., numeric, datetime, categorical, etc.).
is_object_dtype(df["workclass"])

True

### Find the categorical columns with missing values

In [39]:
df.columns

Index(['age', 'workclass', 'education', 'education_num', 'marital_status',
       'occupation', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week',
       'native_country', 'discretized_gross_income'],
      dtype='object')

In [40]:
categorical_columns = [col for col in df.columns if is_object_dtype(df[col])]
categorical_columns

['workclass',
 'education',
 'marital_status',
 'occupation',
 'sex',
 'native_country',
 'discretized_gross_income']

#### Code Explanation

```python
categorical_columns = [col for col in df.columns if is_object_dtype(df[col])]
```

##### 1. `df.columns`

This gives you all the column names of the DataFrame `df`.

---

##### 2. `is_object_dtype(df[col])`

* `df[col]` selects a column from the DataFrame.
* `is_object_dtype(...)` checks if that column has the object data type (usually meaning text/string columns in pandas).

So this returns `True` if the column is categorical/text-like, `False` otherwise.

---

##### 3. List comprehension

```python
[col for col in df.columns if is_object_dtype(df[col])]
```

* Loops through every column in `df.columns`.
* Keeps only those column names where `is_object_dtype(df[col])` is `True`.

Result: a list of all categorical columns (i.e., columns containing string/object data).


---

**In short:**
This code finds all the columns in your DataFrame that are stored as `object` type (usually categorical/text data) and saves their names in the list `categorical_columns`.


In [41]:
for col in categorical_columns:
    print(f"Column: {col}")
    col_values = df[col].unique()
    print(col_values)
    if ("?" in [x.strip() for x in col_values]):
        print("******")
    print(" ")

Column: workclass
[' State-gov' ' Self-emp-not-inc' ' Private' ' Federal-gov' ' Local-gov'
 ' ?' ' Self-emp-inc' ' Without-pay' ' Never-worked']
******
 
Column: education
[' Bachelors' ' HS-grad' ' 11th' ' Masters' ' 9th' ' Some-college'
 ' Assoc-acdm' ' Assoc-voc' ' 7th-8th' ' Doctorate' ' Prof-school'
 ' 5th-6th' ' 10th' ' 1st-4th' ' Preschool' ' 12th']
 
Column: marital_status
[' Never-married' ' Married-civ-spouse' ' Divorced'
 ' Married-spouse-absent' ' Separated' ' Married-AF-spouse' ' Widowed']
 
Column: occupation
[' Adm-clerical' ' Exec-managerial' ' Handlers-cleaners' ' Prof-specialty'
 ' Other-service' ' Sales' ' Craft-repair' ' Transport-moving'
 ' Farming-fishing' ' Machine-op-inspct' ' Tech-support' ' ?'
 ' Protective-serv' ' Armed-Forces' ' Priv-house-serv']
******
 
Column: sex
[' Male' ' Female']
 
Column: native_country
[' United-States' ' Cuba' ' Jamaica' ' India' ' ?' ' Mexico' ' South'
 ' Puerto-Rico' ' Honduras' ' England' ' Canada' ' Germany' ' Iran'
 ' Philippi

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   age                       32561 non-null  int64 
 1   workclass                 32561 non-null  object
 2   education                 32561 non-null  object
 3   education_num             32561 non-null  int64 
 4   marital_status            32561 non-null  object
 5   occupation                32561 non-null  object
 6   sex                       32561 non-null  object
 7   capital_gain              32561 non-null  int64 
 8   capital_loss              32561 non-null  int64 
 9   hours_per_week            32561 non-null  int64 
 10  native_country            32561 non-null  object
 11  discretized_gross_income  32561 non-null  object
dtypes: int64(5), object(7)
memory usage: 3.0+ MB


In [43]:
# replace all the question marks with None
df.replace(' ?' , None, inplace = True)

In [44]:
# Check the number of Nones in each column
df.isnull().sum(axis=0)

age                            0
workclass                   1836
education                      0
education_num                  0
marital_status                 0
occupation                  1843
sex                            0
capital_gain                   0
capital_loss                   0
hours_per_week                 0
native_country               583
discretized_gross_income       0
dtype: int64

#### Code Explanation

##### 1. `df.isnull()`

* This creates a boolean DataFrame of the same shape as `df`.
* Every cell is `True` if the original value is missing (`NaN`), otherwise `False`.

Example (simplified):

| age | workclass |
| --- | --------- |
| 25  | None   |
| np.NaN |  Private        |

→ `df.isnull()` becomes:

| age   | workclass |
| ----- | --------- |
| False | True     |
| True  | False      |

---

##### 2. `.sum(axis=0)`

* When you sum boolean values in pandas, `True` counts as `1` and `False` as `0`.
* `axis=0` means “sum down the rows, for each column.”

So this gives you the count of missing values per column.

---

##### 3. The result

The result is a pandas Series:

* The index is the column names.
* The values are the number of `NaN`s in each column.

---

**In short:**
This code checks for missing values in each column of the DataFrame and returns how many are missing.
For example:

* `workclass` has 1836 missing values.
* `occupation` has 1843 missing values.
* `native_country` has 583  missing values.
* `age`, `education`, and others have none.


In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   age                       32561 non-null  int64 
 1   workclass                 30725 non-null  object
 2   education                 32561 non-null  object
 3   education_num             32561 non-null  int64 
 4   marital_status            32561 non-null  object
 5   occupation                30718 non-null  object
 6   sex                       32561 non-null  object
 7   capital_gain              32561 non-null  int64 
 8   capital_loss              32561 non-null  int64 
 9   hours_per_week            32561 non-null  int64 
 10  native_country            31978 non-null  object
 11  discretized_gross_income  32561 non-null  object
dtypes: int64(5), object(7)
memory usage: 3.0+ MB


In [46]:
df.dropna(axis = 'index', inplace = True)

#### Code Explanation

##### 1. `df.dropna()`

* This function removes missing values **(`NaN` or `None`)** from a DataFrame.
* By default, it drops entire rows if **any** value in that row is missing.

---

##### 2. `axis='index'`

* `axis='index'` (or `axis=0`) means “operate along rows.”
* So: drop rows that contain at least one missing value.
* If you wrote `axis='columns'` (or `axis=1`), it would instead drop entire columns that contain missing values.

---

##### 3. `inplace=True`

* With `inplace=True`, the operation modifies `df` directly and returns `None`.
* Without it (`inplace=False`, the default), it would return a *new DataFrame* while leaving `df` unchanged.

---

##### Effect in practice

If your DataFrame originally had:

|   | age  | workclass | occupation |
| - | ---- | --------- | ---------- |
| 0 | 25.0 | Private   | Tech       |
| 1 | NaN  | Self-emp  | NaN        |
| 2 | 40.0 | NaN       | Sales      |
| 3 | 35.0 | Gov       | NaN        |

After running:

```python
df.dropna(axis='index', inplace=True)
```

the result would be:

|   | age  | workclass | occupation |
| - | ---- | --------- | ---------- |
| 0 | 25.0 | Private   | Tech       |

* Row 0 → no missing values → **kept**.
* Row 1 → missing `age` and `occupation` → **dropped**.
* Row 2 → missing `workclass` → **dropped**.
* Row 3 → missing `occupation` → **dropped**.

---

**Caution:** If you have many missing values, this could shrink your dataset a lot. Sometimes it’s better to fill them (`df.fillna(...)`) or handle them selectively.


In [47]:
# Check the number of Nones in each column
df.isnull().sum(axis=0)

age                         0
workclass                   0
education                   0
education_num               0
marital_status              0
occupation                  0
sex                         0
capital_gain                0
capital_loss                0
hours_per_week              0
native_country              0
discretized_gross_income    0
dtype: int64

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30162 entries, 0 to 32560
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   age                       30162 non-null  int64 
 1   workclass                 30162 non-null  object
 2   education                 30162 non-null  object
 3   education_num             30162 non-null  int64 
 4   marital_status            30162 non-null  object
 5   occupation                30162 non-null  object
 6   sex                       30162 non-null  object
 7   capital_gain              30162 non-null  int64 
 8   capital_loss              30162 non-null  int64 
 9   hours_per_week            30162 non-null  int64 
 10  native_country            30162 non-null  object
 11  discretized_gross_income  30162 non-null  object
dtypes: int64(5), object(7)
memory usage: 3.0+ MB


### 6. Strip off leading spaces in categorical columns.
- In Step 3, when you apply the function unique() to the columns of the DataFrame you observe that the values in categorical columns have a leading space.
- Strip these leading spaces from all values in these columns.

In [49]:
# cycle over the object columns and strip the leading and trailing blanks
for col in categorical_columns:
    print(f"Updating column: {col}")
    df[col] = df[col].str.strip()

Updating column: workclass
Updating column: education
Updating column: marital_status
Updating column: occupation
Updating column: sex
Updating column: native_country
Updating column: discretized_gross_income


In [50]:
df.sample(5)

,age,workclass,education,education_num,marital_status,occupation,sex,capital_gain,capital_loss,hours_per_week,native_country,discretized_gross_income
26949,26,Private,HS-grad,9,Married-civ-spouse,Transport-moving,Male,0,0,40,United-States,>50K
1968,55,Private,10th,6,Married-civ-spouse,Craft-repair,Male,0,0,40,United-States,<=50K
13818,52,Private,HS-grad,9,Separated,Craft-repair,Female,0,0,40,United-States,<=50K
23037,20,Private,HS-grad,9,Never-married,Farming-fishing,Male,0,0,47,United-States,<=50K
29179,20,Private,Some-college,10,Never-married,Other-service,Female,0,0,35,United-States,<=50K


### 7. Replace some values in categorical columns in larger classes. 
This will reduce the number of unique values in these columns and help in creating more readable plots in later milestones.

In Step 3, when you apply the function unique() to the columns of DataFrame, you observe that:
- The column workclass has the unique values 'State-gov’, 'Self-emp-not-inc’, 'Private’, 'Federal-gov’, 'Local-gov’, 'Self-emp-inc’, ‘Without-pay’. Make the following replacements:
    - 'State-gov', 'Federal-gov', 'Local-gov' -> 'Gov' 
    - 'Self-emp-not-inc', 'Self-emp-inc' -> 'Self-Employed'
- The column education has the unique values 'Bachelors’, 'HS-grad’, '11th’, 'Masters’, '9th’, 'Some-college’, 'Assoc-acdm’, '7th-8th’, 'Doctorate’, 'Assoc-voc’, 'Prof-school’, '5th-6th’, '10th’, 'Preschool’, '12th’, ‘1st-4th’. Make the following replacements:
    - '11th’, '9th’, '7th-8th’, '5th-6th’, '10th’, '12th’, ‘1st-4th’ -> ‘School’
- The column marital-status has the unique values 'Never-married’, 'Married-civ-spouse’, 'Divorced’, 'Married-spouse-absent’, 'Separated’, 'Married-AF-spouse’, ‘Widowed’. Make the following replacements:
    - 'Married-civ-spouse’, 'Married-spouse-absent’, ‘Married-AF-spouse’ -> ‘Married’.
- The column occupation has the unique values 'Adm-clerical’, 'Exec-managerial’, 'Handlers-cleaners’, 'Prof-specialty’, 'Other-service’, 'Sales’, 'Transport-moving’, 'Farming-fishing’, 'Machine-op-inspct’, 'Tech-support’, 'Craft-repair’, 'Protective-serv’, 'Armed-Forces’, ‘Priv-house-serv’. Make the following replacements:
    - 'Adm-Clerical' -> 'Clerk', 
    - 'Exec-managerial' -> 'Manager', 
    - 'Prof-specialty' -> 'Professional', 
    - 'Handlers-cleaners', 'Other-service', 'Transport-moving', 'Machine-op-inspct', 'Tech-support', 'Craft-repair', 'Protective-serv', 'Prov-house-serv' -> 'Service', 
    - 'Farming-fishing' -> 'Farming'

In [51]:
# define a dictionary of dictionaries to make the replacements for the columns
# workclass, education, marital_status, and occupation
super_class_dicts = {
    "workclass": {
        'State-gov': 'Gov',
        'Federal-gov': 'Gov',
        'Local-gov': 'Gov',
        'Self-emp-not-inc': 'Self-employed', 
        'Self-emp-inc': 'Self-employed' 
        },
    "education": {
        '11th': 'School', 
        '9th': 'School',
        '7th-8th': 'School', 
        '5th-6th': 'School', 
        '10th': 'School', 
        '12th': 'School', 
        '1st-4th': 'School'
    },
    "marital_status": {
        'Married-civ-spouse': 'Married', 
        'Married-spouse-absent': 'Married', 
        'Married-AF-spouse': 'Married'
    },
    "occupation":{
        'Adm-Clerical': 'Clerk',
        'Exec-managerial': 'Manager',
        'Prof-specialty': 'Professional',
        'Handlers-cleaners': 'Service',
        'Other-service': 'Service',
        'Transport-moving': 'Service', 
        'Machine-op-inspct': 'Service',
        'Tech-support': 'Service',
        'Craft-repair': 'Service', 
        'Protective-serv': 'Service', 
        'Prov-house-serv': 'Service',
        'Farming-fishing': 'Farming'
    }
}

In [52]:
# Create new superclass columns
for col in ["workclass", "education","marital_status","occupation"]:
    col_sc = col+"_sc"
    dict_sc = super_class_dicts[col]
    df[col_sc] = df[col].apply(lambda x: x if x not in dict_sc else dict_sc[x])
    print(f"\nCreated column {col_sc}")
    print(df.groupby([col,col_sc]).size().to_frame(name="row_count").sort_index())


Created column workclass_sc
                                row_count
workclass        workclass_sc            
Federal-gov      Gov                  943
Local-gov        Gov                 2067
Private          Private            22286
Self-emp-inc     Self-employed       1074
Self-emp-not-inc Self-employed       2499
State-gov        Gov                 1279
Without-pay      Without-pay           14

Created column education_sc
                           row_count
education    education_sc           
10th         School              820
11th         School             1048
12th         School              377
1st-4th      School              151
5th-6th      School              288
7th-8th      School              557
9th          School              455
Assoc-acdm   Assoc-acdm         1008
Assoc-voc    Assoc-voc          1307
Bachelors    Bachelors          5044
Doctorate    Doctorate           375
HS-grad      HS-grad            9840
Masters      Masters            1627
Preschool

#### Code Explanation

##### 1. `super_class_dicts`

This is a dictionary of dictionaries.

* Each *key* is a column name (`workclass`, `education`, `marital_status`, `occupation`).
* Each *value* is another dictionary that maps fine-grained categories to broader "super classes".

Example:

```python
"workclass": {
    'State-gov': 'Gov',
    'Federal-gov': 'Gov',
    ...
}
```

So `"State-gov"` → `"Gov"`.

---

##### 2. The loop

```python
for col in ["workclass", "education","marital_status","occupation"]:
```

This iterates over the four categorical columns that need grouping.

---

##### 3. Create new column names

```python
col_sc = col+"_sc"
```

* For each column, it makes a new one with `_sc` (short for *super class*) appended.

  * `workclass` → `workclass_sc`
  * `education` → `education_sc`
  * etc.

---

##### 4. Get the right mapping dictionary

```python
dict_sc = super_class_dicts[col]
```

This extracts the dictionary of mappings for the current column.

---

##### 5. Apply transformation

```python
df[col_sc] = df[col].apply(lambda x: x if x not in dict_sc else dict_sc[x])
```

* For each row, look at the value `x` in the column.
* If `x` is a key in `dict_sc`, replace it with its mapped value.
* Otherwise, keep `x` unchanged.

Example:

* `df["workclass"] = "Federal-gov"` → `"Gov"`
* `df["workclass"] = "Private"` → stays `"Private"` (since not in dictionary).

---

##### 6. Print results

```python
print(f"\nCreated column {col_sc}")
print(df.groupby([col,col_sc]).size().to_frame(name="row_count").sort_index())
```

* This prints the new column name created.
* Then groups by original value and its corresponding super class, counting rows.

---

##### Result explained

Example for **workclass**:

```
workclass        workclass_sc        
Federal-gov      Gov              943
Local-gov        Gov             2067
Private          Private        22286
Self-emp-inc     Self-employed   1074
Self-emp-not-inc Self-employed   2499
State-gov        Gov             1279
Without-pay      Without-pay       14
```

* `"Federal-gov"`, `"Local-gov"`, `"State-gov"` are all mapped to `"Gov"`.
* `"Self-emp-inc"` and `"Self-emp-not-inc"` are mapped to `"Self-employed"`.
* Categories not in the mapping (like `"Private"`, `"Without-pay"`) stay the same.
* The counts (`943`, `2067`, …) show how many rows belong to each original category → super class.

Similar logic applies to `education`, `marital_status`, and `occupation`.

---

**In short:**
This code creates *simplified categories* (super classes) for multiple categorical columns in a DataFrame, by grouping detailed values (like `"Federal-gov"`, `"State-gov"`) into broader ones (like `"Gov"`). Then it shows how many rows fall into each original → super class mapping.



In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30162 entries, 0 to 32560
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   age                       30162 non-null  int64 
 1   workclass                 30162 non-null  object
 2   education                 30162 non-null  object
 3   education_num             30162 non-null  int64 
 4   marital_status            30162 non-null  object
 5   occupation                30162 non-null  object
 6   sex                       30162 non-null  object
 7   capital_gain              30162 non-null  int64 
 8   capital_loss              30162 non-null  int64 
 9   hours_per_week            30162 non-null  int64 
 10  native_country            30162 non-null  object
 11  discretized_gross_income  30162 non-null  object
 12  workclass_sc              30162 non-null  object
 13  education_sc              30162 non-null  object
 14  marital_status_sc         3

In [54]:
df.to_csv(data_loc+"cleaned_adult_consensus.csv", index=False)